### PR TITLE
🧹 simplify read examples

### PR DESCRIPTION
**Pull request by St0rmBr3w**

### CODE REVIEW

File: examples/view-pure-read/package.json [hunk 1]
```diff
@@ -0,0 +1,73 @@
+{
+  "name": "@layerzerolabs/view-pure-read-example",
+  "version": "0.2.0",
```
Comment (janjakubnanista) [inline examples/view-pure-read/package.json:3]: `0.0.1`

File: examples/view-pure-read/CHANGELOG.md
Comment (janjakubnanista) [inline examples/view-pure-read/CHANGELOG.md:1]: Please delete

**Review by janjakubnanista — CHANGES_REQUESTED** [review]
Just some small stuff

File: examples/view-pure-read/package.json [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  },
+  "devDependencies": {
+    "@babel/core": "^7.23.9",
+    "@layerzerolabs/eslint-config-next": "~2.3.39",
+    "@layerzerolabs/lz-definitions": "^3.0.12",
+    "@layerzerolabs/lz-evm-messagelib-v2": "^3.0.12",
+    "@layerzerolabs/lz-evm-protocol-v2": "^3.0.12",
+    "@layerzerolabs/lz-evm-v1-0.7": "^3.0.12",
+    "@layerzerolabs/lz-v2-utilities": "^3.0.12",
+    "@layerzerolabs/oapp-evm": "^0.2.0",
```
Comment (janjakubnanista) [inline examples/view-pure-read/package.json:31]: This version is outdated, should be 0.3.0

**Review by janjakubnanista — DISMISSED** [review]

File: packages/oapp-evm-upgradeable/artifacts/ExecutorOptions.sol/ExecutorOptions.json [hunk 3]
```diff
@@ -5,6 +5,7 @@
       "name": "Executor_InvalidLzComposeOption",
       "inputs": []
     },
+    { "type": "error", "name": "Executor_InvalidLzReadOption", "inputs": [] },
```
[the remark this answers is not in the document]
Comment (St0rmBr3w) [reply packages/oapp-evm-upgradeable/artifacts/ExecutorOptions.sol/ExecutorOptions.json:8]: I believe so, the artifact is just from the fact that ExecutorOptions have been changed protocol wide (so that there is now an `ExecutorLzReadOption`:

https://github.com/LayerZero-Labs/LayerZero-v2/blob/943ce4a2bbac070f838e12c7fd034bca6a281ccf/packages/layerzero-v2/evm/messagelib/contracts/libs/ExecutorOptions.sol#L21

Believe this is just being picked up in the artifacts now when running `pnpm build` @ryandgoulding.

File: examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        // Verify that targetContractAddress is correctly set
+        assertEq(bOApp.targetContractAddress(), address(exampleContract));
+    }
+
+    /**
+     * @notice Tests sending a read request and handling the received sum.
+     *
+     * @dev Simulates a user initiating a read request to add two numbers and verifies that the SumReceived event is emitted with the correct sum.
+     */
+    function test_send_read() public {
```
[the remark this answers is not in the document]
Comment (St0rmBr3w) [reply examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol:113]: This is a good callout, will play around with fuzzing the input params.

File: examples/oapp-read/test/foundry/ReadPublic.t.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+import "forge-std/console.sol";
+import "forge-std/Test.sol";
+
+// DevTools imports
+import { TestHelperOz5 } from "@layerzerolabs/test-devtools-evm-foundry/contracts/TestHelperOz5.sol";
+
+contract ReadPublicTest is TestHelperOz5 {
+    using OptionsBuilder for bytes;
+
+    uint32 private aEid = 1; // Chain A Endpoint ID
```
[the remark this answers is not in the document]
Comment (St0rmBr3w) [reply examples/oapp-read/test/foundry/ReadPublic.t.sol:27]: Should this be made consistent across all our unit tests?

I believe this is inconsistent in other `./examples`:

`./examples/oapp`: https://github.com/LayerZero-Labs/devtools/blob/9c95650097c17113b4bb6128ce5479f68eb55acd/examples/oapp/test/foundry/MyOApp.t.sol#L23-L24

`./examples/oft`: https://github.com/LayerZero-Labs/devtools/blob/9c95650097c17113b4bb6128ce5479f68eb55acd/examples/oft/test/foundry/MyOFT.t.sol#L31-L32

**Comment by ryandgoulding** [comment]
Fixed pnpm-lock.yaml.  Running through CI pipeline.

**Comment by nazreen** [comment]
This PR is awesome. when can we have this in main ?

File: examples/oapp-read/contracts/ReadPublic.sol [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+     * @param _delegate The address that will have ownership privileges.
+     * @param _readChannel The LayerZero read channel ID.
+     */
+    constructor(
+        address _endpoint,
+        address _delegate,
+        uint32 _readChannel
+    ) OAppRead(_endpoint, _delegate) Ownable(_delegate) {
+        READ_CHANNEL = _readChannel;
+        _setPeer(_readChannel, AddressCast.toBytes32(address(this)));
```
Comment (shankars99) [inline examples/oapp-read/contracts/ReadPublic.sol:45]: why doesnt `OAppRead` set self peer? in lzRead we always self peer dont we?

File: examples/oapp-read/contracts/ReadPublic.sol
Comment (shankars99) [inline examples/oapp-read/contracts/ReadPublic.sol:1]: can we have an example that switches between `lzRead` and `lzSend`?

File: examples/oapp-read/test/foundry/ReadPublic.t.sol [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    function test_constructor() public {
+        assertEq(bOApp.owner(), address(this));
+        assertEq(address(bOApp.endpoint()), address(endpoints[bEid]));
+    }
+
+    function test_send_read() public {
+        bytes memory options = OptionsBuilder.newOptions().addExecutorLzReadOption(1e8, 100, 0);
+        console.logBytes(options);
```
Comment (shankars99) [inline examples/oapp-read/test/foundry/ReadPublic.t.sol:75]: remove the `console.log` ?

File: examples/oapp-read/test/foundry/ReadPublic.t.sol [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        uint256 dataReceived;
+        for (uint256 i = 0; i < entries.length; i++) {
+            Vm.Log memory entry = entries[i];
+            if (entry.topics[0] == keccak256("DataReceived(uint256)")) {
+                dataReceived = abi.decode(entry.data, (uint256));
+                found = true;
+                break;
+            }
+        }
+        require(found, "DataReceived event not found");
```
Comment (shankars99) [inline examples/oapp-read/test/foundry/ReadPublic.t.sol:106]: nit: use assert instead of require so that we're using foundry Test stuff in the foundry test

File: examples/view-pure-read/contracts/ReadViewOrPure.sol [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    uint32 public READ_CHANNEL;
+
+    /// @notice Message type for the read operation.
+    uint16 public constant READ_TYPE = 1;
+
+    /// @notice Target chain's Endpoint ID.
+    uint32 public targetEid;
+
+    /// @notice Target contract address on the target chain.
+    address public targetContractAddress;
```
Comment (shankars99) [inline examples/view-pure-read/contracts/ReadViewOrPure.sol:37]: can be `immutable` since the values are not changed

File: examples/view-pure-read/contracts/ReadViewOrPureAndCompute.sol [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    uint32 public READ_CHANNEL;
+
+    /// @notice Message type for the read operation.
+    uint16 public constant READ_TYPE = 1;
+
+    /// @notice Target chain's Endpoint ID.
+    uint32 public targetEid;
+
+    /// @notice Target contract address on the target chain.
+    address public targetContractAddress;
```
Comment (shankars99) [inline examples/view-pure-read/contracts/ReadViewOrPureAndCompute.sol:41]: `immutable` since the value isn't changed

File: examples/view-pure-read/foundry.toml [hunk 11]
```diff
@@ -0,0 +1,27 @@
+[profile.default]
+solc-version = '0.8.22'
+src = 'contracts'
+out = 'out'
+test = 'test/foundry'
+cache_path = 'cache/foundry'
+libs = [
```
Comment (shankars99) [inline examples/view-pure-read/foundry.toml:7]: add in 

```
optimizer = true
optimizer_runs = 1_000
```

so that `forge-stable` does not complain about stack too deep

File: examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        uint256 sumReceived;
+        for (uint256 i = 0; i < entries.length; i++) {
+            Vm.Log memory entry = entries[i];
+            if (entry.topics[0] == keccak256("SumReceived(uint256)")) {
+                sumReceived = abi.decode(entry.data, (uint256));
+                found = true;
+                break;
+            }
+        }
+        require(found, "SumReceived event not found");
```
Comment (shankars99) [inline examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol:153]: use `assert` which is a foundry Test function and not require which is standard solidity stuff

File: examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+        vm.prank(userA);
+        bOApp.readSum{ value: fee.nativeFee }(a, b, options);
+
+        // Simulate processing the response packet to bOApp on bEid, injecting the sum
+        this.verifyPackets(
+            bEid,
+            addressToBytes32(address(bOApp)),
+            0,
+            address(0x0),
+            abi.encode(a + b) // The sum of a and b
```
Comment (shankars99) [inline examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol:138]: fuzz here because `a+b` could overflow

File: examples/view-pure-read/test/foundry/ReadViewOrPureAndCompute.t.sol [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        uint256 sumReceived;
+        for (uint256 i = 0; i < entries.length; i++) {
+            Vm.Log memory entry = entries[i];
+            if (entry.topics[0] == keccak256("SumReceived(uint256)")) {
+                sumReceived = abi.decode(entry.data, (uint256));
+                found = true;
+                break;
+            }
+        }
+        require(found, "SumReceived event not found");
```
Comment (shankars99) [inline examples/view-pure-read/test/foundry/ReadViewOrPureAndCompute.t.sol:168]: `assert` over `require`

File: examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol (hunk 13 above)
Comment (St0rmBr3w) [reply examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol:138]: Think this is unnecessary since this is an example contract, there will likely be edge cases that cause a + b to overflow.

**Review by shankars99 — DISMISSED** [review]
lgtm!

File: examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol (hunk 13 above)
Comment (shankars99) [reply examples/view-pure-read/test/foundry/ReadViewOrPure.t.sol:138]: yeah its alright tbh

File: examples/view-pure-read/contracts/ReadViewOrPure.sol [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+    /**
+     * @notice Combines the options for messaging.
+     *
+     * @param _channelId The channel ID.
+     * @param _msgType The message type.
+     * @param _extraOptions Additional options.
+     *
+     * @return options The combined options.
+     */
+    function _combineOptions(
```
Comment (ryandgoulding) [inline examples/view-pure-read/contracts/ReadViewOrPure.sol:173]: Why is this needed v.s. just inlining calls to OAppOptionsType3.combineOptions(...)?

File: examples/view-pure-read/contracts/ReadViewOrPureAndCompute.sol [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+     * @param _extraOptions Additional options.
+     *
+     * @return options The combined options.
+     */
+    function _combineOptions(
+        uint32 _channelId,
+        uint16 _msgType,
+        bytes calldata _extraOptions
+    ) internal view returns (bytes memory options) {
+        options = combineOptions(_channelId, _msgType, _extraOptions);
```
Comment (ryandgoulding) [inline examples/view-pure-read/contracts/ReadViewOrPureAndCompute.sol:231]: same here?

**Comment by ryandgoulding** [comment]
https://github.com/LayerZero-Labs/devtools/pull/1338
